### ISSUE TITLE
Adding compressors to etmoses gas asset calculation 

### ISSUE THREAD

**Issue by grdw** [quintel/etmoses#960]
There's a new type of gas asset in ETMoses: the compressor. It's purpose is to compress gas from a lower pressure level to a higher pressure level. The current `Upward` capacity is Infinite but this should be set to a calculated amount based on the compressors.

**Comment by grdw** [comment]
I'm looking at the code directly but I can stare at it for days and probably just barely touch the essence of what is going on; also reading the specs doesn't make much sense due to the level of abstraction. So therefor these questions:

"The compressors need to control the `Upward` capacity of the Gas Load"
I was told this but I don't really have a clue what that means. How exactly does that influence the Gas Load? I understand that the compressor compresses gas so it's transportable (moveable) from for instance 4 bar to 8 bar. Ofcourse there's some energy lost (therefor the efficiency) but what has capacity got to do with it? Is that how much gas it can compress? I'm a bit lost here... I understand what it does but I have no idea how.

I was looking at `Slots` which are two pieces of `Slot` and I see that in there the connectors are the only things used from the `@assets`. My first little gut-feeling tells me that the compressors also have to be incorporated in here somewhere but than how would I correctly put this into specs? 

To expand. What will happen right now is: I'll change `gas_chain_spec.rb`, just add a compressor and wait for an efficiency to differt from it's normal behavior and change the expectation accordingly blindly assuming it's correct. Which is not really something I should be doing.

--

Long story short: how does a compressor work and how does it influence the Gas load calculation?

**Comment by antw** [comment]
I'm confused. Isn't upward efficiency and capacity already modelled in `Connectors` using [the `low_to_high` capacity and efficiency](https://github.com/quintel/etmoses/blob/71e320c2a5b62e22b5ed58bc9bb1ec15589b20f3/db/static/connectors/inefficient_connector.yml#L1-L6)?

**Comment by ChaelKruip** [comment]
> Is that how much gas it can compress?

Yes. If more gas is being pushed upward then the compressor can handle, the gas is 'flared' away.

> (explain it like I'm 5)

OK, a long, long time ago there was a beautiful princess in a very big and cold castle. She asked all the young princes from neighbouring countries to make her feel warm again. Her evil stepmother didn't like the sound of that at all and gave her a little gas-fired heater instead so she could warm up her crystal room high up in the highest tower of the castle. Because the tower was so high, the gas would not go up there out of its free will. The king installed a compressor to pump the gas upward to be used in the gas-fired heater. 
Het heater needed 2kW of gas and the king fed those 2kW into the compressor but the poor compressor only had a capacity of 1kW. It pumped and it pumped its little lungs full of gas but only half of the 2kW made it to the princess. The other 1kW was coughed up by the brave little compressor (who is the actual hero of our story) filling up the dungeons of the castle with an explosive mixture. The king, content with his solution lit a nice pipe and the castle deflagrated in a ball of fire. Nobody survived the tragedy but the scorching heat was able to melt the frozen tears of the icy princess. So, that was somewhat of a happy ending right?

**Comment by ChaelKruip** [comment]
> Isn't upward efficiency and capacity already modelled in Connectors using the low_to_high capacity and efficiency?

Yes, because we (I) thought at the time that the compressor would be an integral part of the connectors (GOS, DRS etc.). It turns out that they are not... But I think every *connection between two levels* still has a `low_to_high` capacity. Does that make sense @antw?

**Comment by grdw** [comment]
> So, that was somewhat of a happy ending right?

It needs a bit more dragons but I get it though. 😄 

> I'm confused. 

So am I.

---

So the `low_to_high` capacity of the compressor is *added* to the `low_to_high` capacity of the connector? Or is one swapped out for the other?

**Comment by ChaelKruip** [comment]
> So the low_to_high capacity of the compressor is added to the low_to_high capacity of the connector? Or is one swapped out for the other?

I think of it like this: the *connection* between level 1 and level 2 (connection 1 <--> 2) has two directions of possible gas-flow. The connection is not a single entity but rather the effective sum of all connectors that are in the physical network (GOS, DRS etc.)

The direction from high to low pressure is simple at the moment in the sense that it is of limitless (infinite) capacity and lossless. If we can find some numbers (capacities) for the different (physical) stations that *add up to the total connection between two levels* we can add up the capacities of the stations to yield the total (finite) capacity for the high_to_low gas-flow (see https://github.com/quintel/etmoses/issues/861#issuecomment-199861250). 

For the direction from low to high, the capacity and the losses are determined by the sum of the capacities and the weighted sum of the efficiencies of the compressors that are part of the connection. 

Summarising: the total (unphysical) connection between two levels has two directions. The high_to_low is set by GOS, DRS etc. and is finished. The low_to_high capacity and efficiency needs to be set by the compressors.

**Comment by grdw** [comment]
> I think of it like this: the connection between level 1 and level 2 (connection 1 <--> 2) has two directions of possible gas-flow. The connection is not a single entity but rather the effective sum of all connectors that are in the physical network (GOS, DRS etc.)

It currently is.

> The low_to_high capacity and efficiency needs to be set by the compressors.

Ignoring the `low_to_high` capacity of all the connectors... ?

**Comment by ChaelKruip** [comment]
> Ignoring the low_to_high capacity of all the connectors... ?

If by connectors you mean the GOS, DRS etc. things then yes. These units turn out not to have a low_to_high anything...

**Comment by grdw** [comment]
> These units turn out not to have a low_to_high anything...

Aha, that makes sense. Thanks.

**Comment by ChaelKruip** [comment]
What is the status @grdw ?

**Comment by antw** [comment]
> What is the status @grdw ?

There's a pull request (#965). I understood it was awaiting "[ - ] Test it in a real LES", but perhaps that has been completed? It seems there is data in place from @jorisberkhout; assuming that data is ready, and the testing finished, I can work on getting the PR merged.

> [Connectors] turn out not to have a low_to_high anything...

Is it the case that connectors have no `low_to_high` anything, and compressors have no `high_to_low` anything? Meaning: connectors do high-pressure to lower-pressure exclusively, and compressors do only low-pressure to higher-pressure?

Is there any case that you know of where either of these technologies might have both high-to-low and low-to-high attributes?

I ask because the relevant pull request still uses `high_to_low` and `low_to_high` attributes, but if each technology is only one-way we can probably simplify this.

**Comment by ChaelKruip** [comment]
> connectors do high-pressure to lower-pressure exclusively, and compressors do only low-pressure to higher-pressure?

Yes!

> Is there any case that you know of where either of these technologies might have both high-to-low and low-to-high attributes?

No. Not at the moment.

>  if each technology is only one-way we can probably simplify this.

👍

**Comment by ChaelKruip** [comment]
> assuming that data is ready, and the testing finished, I can work on getting the PR merged.

Good to go!
